### DISCUSSION TITLE
Rounded corners not working?

### DISCUSSION THREAD

**Discussion by kolunio82** [valinet/ExplorerPatcher#291] [Q&A]
hi, i downloaded lastest version and no rounded corners not working at all. Even after restart. I need to do something extra?

**Comment by kolunio82** [comment]
In latest also don't work for me.

**Reply by valinet** [reply]
EP has no such functionality at the moment. That other utility I made is separate, and idk, maybe it broke on the latest development build and needs fixing, other than that, it is in no way tied to new EP releases...

**Comment by kolunio82** [comment]
https://github.com/valinet/Win11DisableRoundedCorners/releases

when i use this i have message unable to restore dwn

**Reply by valinet** [reply]
Then restore manually... you can also patch manually... Thanks

**Reply by kolunio82** [reply]
> 
> 
> Then restore manually... you can also patch manually... Thanks

can you tell me how?

@ECHO OFF
taskkill /F /IM dwm.exe
taskkill /F /IM conhost.exe

whe i use this it starts again with no changed corners look

**Reply by valinet** [reply]
Look at the source code, please, it tells pretty clearly what to do manually. 

First of all, ALL numbers I am talking about are in base 16 aka hexadecimal. We always talk in base 16, just to clear any confusion. Maths works the same no matter the numeration base.

The idea is this: identify the location of the `CTopLevelWindow::GetEffectiveCornerStyle` function in `uDWM.dll`. For that, download its symbol file using [PDBDownloader.exe](https://github.com/rajkumar-rangaraj/PDB-Downloader/releases/download/v1.0/PDBDownloader.exe) and grab the symbol location using [pdbdump.zip](https://github.com/valinet/ExplorerPatcher/files/7402969/pdbdump.zip) for example: `pdbdump.exe uDWM.pdb | findstr "GetEffectiveCornerStyle`. From that, you substract the base address (which is usually `0x400000`, mentioned in the source code of pdbdump [here](https://gist.github.com/phaufe/2973037) along with other relevant info), for me that command returns `0000000000426ee4` for `uDWM.dll` from 22000.318, so the offset I am interested into is `0x26ee4` (`= 0x426ee4 - 0x400000`). 

You then open the file in some hex editor, like [HxD](https://mh-nexus.de/downloads/HxDSetup.zip), go to that offset from the beginning of the file, that is where that function begins. You replace the bytes there with whatever bytes you want, which represent the instructions you want executed instead. Again, take that from the source code, for example, type this `48 C7 C0 00 00 00 00 c3`, which means `mov rax, 0; ret`. "`rax`" contains the thing a function returns on amd64 architecture. To assemble/disassemble op codes quickly, I recommend [this site](https://defuse.ca/online-x86-assembler.htm).

So, in higher language, the way this works, is that when asked about the "effective corner style", `uDWM` always answers 0, which by experimentation I determined to mean "not rounded corners".

That's it, with that, you should be good to go, please try to think logically at every step, the provided info is more than enough for anyone to pull this off, try to figure out basic problems yourself first. If it's something you absolutely cannot get past, sure, ask again.

Of course, if one's not careful and mindful and do not pay attention at every step and skip understanding some of them, one will fuck up the patch and the system will become pretty borked since there is no program to show you anything on the screen anymore. Thus, my recommendation is to attempt this in a VM first, see that it works just fine there, and then move the patch to the physical machine where it will work just as fine as well. The patch is simple, if done right, it will work 100%.

Thanks

**Reply by kolunio82** [reply]
I will try. It's not easy for me due to my palms disability. thanks

**Reply by valinet** [reply]
Pffff... sorry to hear that. The tool should work mainly, idk why it sometimes fails. It's a computer program, sometimes there are bugs; I listed the instructions so that you can reproduce manually, other than that, idk how else to help. Thank you as well.

**Reply by kolunio82** [reply]
Maybe I do something wrong? Before latest winows updates program works. Now when my no rounder corners gone I tried once again and I see this mssage, even when I run program in admin mode.

**Reply by valinet** [reply]
I don't know, I cannot guess. Again, try to debug it. The way it works is described above. It's just a tool I made for fun/proof of concept etc. I cannot fix it on every update or provide support. The knowledge is all available, try to see what broke it and why it does not work anymore manually. Again, I am not a magician, I cannot make guesses on matters like these. Thanks.

**Reply by kolunio82** [reply]
I don't know how to debug it. I'm sorry but all is difficult with my palms. Maybe we can connect by anydesk if you will have a moment to help me?

**Reply by valinet** [reply]
I don't know, maybe something changed in the latest updates... if we connect and I copy a broken file, we would lose image and that's be it. It's risky. And for the life in me I cannot understand why is this so important. If you depend on the computer to work, and also considering your condition, why not stop messing with the OS and leave it the way it was shipped by the OS manufacturer which is tested and stable and works...?

**Reply by kolunio82** [reply]
becouse not rounded corners are much easier to use for me.

**Reply by cableghost** [reply]
I installed EP and I still have rounded corners. I did not find a specific option in Setup. Is this an automatic change or something?

I did also try the standalone approach, but could not get it to extract what was needed.

Win 11 build: 22000.469

**Reply by valinet** [reply]
~~In EP, go to Updates, check "Receieve pre-release versions", click "Update program now and restart File Explorer". After the update is installed and it restarts, you'll fine the option under the "Other" section. Sorry I forgot to mention this in the earlier post, this functionality is not yet available in the release for EP, only in pre-releases. It will roll out soon to all users as a release. That's also the reason why the `extract` command did not extract it for you, as it's simply not there in the version labeled as "release". After you upgrade, if you wish to only receive releases as updates, don't forget to uncheck the option I mentioned. You can find `ep_dwm.exe` if you need it in `C:\Program Files\ExplorerPatcher` if needed, or you can use `ep_setup /extract` but on a newer `ep_setup`. Thanks.~~

**Reply by valinet** [reply]
Actually, I have just promoted `22000.469.41.16` as a release, just check for updates regularly in EP and it will download and install that, and that has the option to disable rounded corners, no need to roll with pre-releases. Thanks.

**Reply by cableghost** [reply]
Thank you, though I cannot reopen the app. I've uninstalled, re-installed new release and it will not open. Upon clicking on the exe file from extracting the setup file, I get nothing. Please advise.

**Reply by valinet** [reply]
I don't understand. You tried running `ep_dwm.exe` directly? Yeah, it won't work. It has to be registered as a system service, as described at https://github.com/valinet/ep_dwm. It cannot run like this, it has to run as SYSTEM account in order to be able to modify `dwm.exe`'s memory.
